src/main/java/com/aws/greengrass/clientdevices/auth/iot/registry/ThingRegistry.java

```diff
@@ -5,33 +5,101 @@
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import javax.inject.Inject;
 
 public class ThingRegistry {
-    // holds mapping of thingName to IoT Certificate ID;
+    // holds mapping of thingName to IoT Certificate IDs;
     // size-bound by default cache size, evicts oldest written entry if the max size is reached
-    private final Map<String, String> registry = Collections.synchronizedMap(
-            new LinkedHashMap<String, String>(RegistryConfig.REGISTRY_CACHE_SIZE, 0.75f, false) {
+    private final Map<String, List<String>> registry = Collections.synchronizedMap(
+            new LinkedHashMap<String, List<String>>(RegistryConfig.REGISTRY_CACHE_SIZE, 0.75f, false) {
                 @Override
                 protected boolean removeEldestEntry(Map.Entry eldest) {
                     return size() > RegistryConfig.REGISTRY_CACHE_SIZE;
                 }
             });
 
     private final IotAuthClient iotAuthClient;
+    private final DomainEvents domainEvents;
 
     @Inject
-    public ThingRegistry(IotAuthClient iotAuthClient) {
+    public ThingRegistry(IotAuthClient iotAuthClient, DomainEvents domainEvents) {
         this.iotAuthClient = iotAuthClient;
+        this.domainEvents = domainEvents;
+    }
+
+    /**
+     * Retrieve a Thing based on ThingName.
+     *
+     * @param thingName ThingName
+     * @return Thing domain object
+     */
+    public Thing getThing(String thingName) {
+        List<String> certificateIds = registry.get(thingName);
+        if (certificateIds != null) {
+            return new Thing(thingName, certificateIds);
+        }
+        return null;
+    }
+
+    /**
+     * Attach a certificate to Thing.
+     * </p>
+     * If the provided Thing does not exist, it will be created.
+     * @param thingName     ThingName.
+     * @param certificateId Certificate ID to attach.
+     */
+    public synchronized void attachCertificateToThing(String thingName, String certificateId) {
+        List<String> certificateIds = registry.get(thingName);
+
+        if (certificateIds != null && certificateIds.contains(certificateId)) {
+            // Nothing to do
+            return;
+        }
+
+        // Thing doesn't exist - create it
+        if (certificateIds == null) {
+            certificateIds = new ArrayList<>(Collections.singletonList(certificateId));
+            registry.put(thingName, certificateIds);
+        } else {
+            certificateIds.add(certificateId);
+        }
+
+        domainEvents.emit(
+                new ThingEvent(ThingEvent.ThingEventType.THING_UPDATED,
+                        thingName,
+                        new ArrayList<>(certificateIds)));
+    }
+
+    /**
+     * Detach a certificate from Thing.
+     * @param thingName     ThingName.
+     * @param certificateId Certificate ID to detach.
+     */
+    public synchronized void detachCertificateFromThing(String thingName, String certificateId) {
```

Comment:
Similar to my comment above, if we were operating on `Things` this method would not exist we would just update the Thing and call the registry's `update` or `save` method. Also would remove the duplication in event emitting